plugin/src/main/java/net/aufdemrand/denizen/objects/dPlayer.java

```diff
@@ -1702,6 +1703,57 @@ else if (isOnline()) {
                     .getAttribute(attribute.fulfill(1));
         }
 
+        if (attribute.startsWith("attack_cooldown")) {
+            attribute.fulfill(1);
+
+            if (!NMSHandler.getVersion().isAtLeast(NMSVersion.v1_9_R2)) {
+                dB.echoError("Attack cooldowns don't exist prior to 1.9.");
+                return null;
+            }
+
+            // <--[tag]
+            // @attribute <p@player.attack_cooldown.time_passed>
+            // @returns Duration
+            // @description
+            // Returns the amount of time that passed since the start of the attack cooldown.
+            // -->
+            if (attribute.startsWith("time_passed")) {
+                return new Duration((long) NMSHandler.getInstance().getPlayerHelper()
+                        .ticksPassedDuringCooldown(getPlayerEntity())).getAttribute(attribute.fulfill(1));
+            }
+
+
+            // <--[tag]
+            // @attribute <p@player.attack_cooldown.max_duration>
+            // @returns Duration
+            // @description
+            // Returns the maximum amount of time that can pass before the player's main hand has returned
+            // to its original place after the cooldown has ended.
+            // NOTE: This is slightly inaccurate and may not necessarily match with the actual attack
+            // cooldown progress.
+            // -->
+            else if (attribute.startsWith("max_duration")) {
+                return new Duration((long) NMSHandler.getInstance().getPlayerHelper()
+                        .getMaxAttackCooldownTicks(getPlayerEntity())).getAttribute(attribute.fulfill(1));
+            }
+
+
+            // <--[tag]
+            // @attribute <p@player.attack_cooldown.percent_done>
+            // @returns Element(Number)
+            // @description
+            // Returns the progress of the attack cooldown. 0 means that the attack cooldown has just
+            // started, while 100 means that the attack cooldown has finished.
+            // NOTE: This may not match exactly with the clientside attack cooldown indicator.
+            // -->
+            else if (attribute.startsWith("percent_done")) {
+                return new Element(NMSHandler.getInstance().getPlayerHelper()
+                        .getAttackCooldownPercent(getPlayerEntity()) * 100).getAttribute(attribute.fulfill(1));
+            }
+
+            return null;

```

Comment:
This should probably include an error here ^ since you've fulfilled an attribute but processed nothing

Reply:
I'm not sure what the error message would say. Would `"The attack_cooldown tag shouldn't be used alone!"` be enough?

Reply:
Probably: `"The tag 'player.attack_cooldown...' must be followed by a sub-tag."` or something.

The other option is to avoid fulfilling an attribute when reaching this spot, or just not having subtags at all (instead have `player.attack_cooldown_percentage` and such)